metricbeat/mb/module/wrapper.go

```diff
@@ -182,128 +184,155 @@ func (mw *Wrapper) Hash() uint64 {
 
-func (msw *metricSetWrapper) singleEventFetch(fetcher mb.EventFetcher) (common.MapStr, error) {
-	start := time.Now()
+func (msw *metricSetWrapper) singleEventFetch(fetcher mb.EventFetcher, reporter *eventReporter) {
+	reporter.startFetchTimer()
 	event, err := fetcher.Fetch()
-	elapsed := time.Since(start)
+	reporter.ErrorWith(err, event)
+}
 
-	if err == nil {
-		msw.stats.success.Add(1)
-	} else {
-		msw.stats.failures.Add(1)
+func (msw *metricSetWrapper) multiEventFetch(fetcher mb.EventsFetcher, reporter *eventReporter) {
+	reporter.startFetchTimer()
+	events, err := fetcher.Fetch()
+	for _, event := range events {
+		reporter.ErrorWith(err, event)
 	}
+}
+
+func (msw *metricSetWrapper) reportingFetch(fetcher mb.ReportingFetcher, reporter *eventReporter) {
+	reporter.startFetchTimer()
+	fetcher.Fetch(reporter)
+}
 
-	if event, err = createEvent(msw, event, err, start, elapsed); err != nil {
-		return nil, errors.Wrap(err, "createEvent failed")
+// close closes the underlying MetricSet if it implements the mb.Closer
+// interface.
+func (msw *metricSetWrapper) close() error {
+	if closer, ok := msw.MetricSet.(mb.Closer); ok {
+		return closer.Close()
 	}
+	return nil
+}
 
-	return event, nil
+// String returns a string representation of metricSetWrapper.
+func (msw *metricSetWrapper) String() string {
+	return fmt.Sprintf("metricSetWrapper[module=%s, name=%s, host=%s]",
+		msw.module.Name(), msw.Name(), msw.Host())
 }
 
-func (msw *metricSetWrapper) multiEventFetch(fetcher mb.EventsFetcher) ([]common.MapStr, error) {
-	start := time.Now()
-	events, err := fetcher.Fetch()
-	elapsed := time.Since(start)
+// Reporter implementation
 
-	var rtnEvents []common.MapStr
-	if err == nil {
-		msw.stats.success.Add(1)
+// eventReporter implements the Reporter interface which is a callback interface
+// used by MetricSet implementations to report an event(s), an error, or an error
+// with some additional metadata.
+type eventReporter struct {
+	msw   *metricSetWrapper
+	done  <-chan struct{}
+	out   chan<- common.MapStr
+	start time.Time // Start time of the current fetch (or zero for push sources).
+}
 
-		for _, event := range events {
-			if event, err = createEvent(msw, event, nil, start, elapsed); err != nil {
-				return nil, errors.Wrap(err, "createEvent failed")
-			}
-			if event != nil {
-				rtnEvents = append(rtnEvents, event)
-			}
-		}
+// startFetchTimer demarcates the start of a new fetch. The elapsed time of a
+// fetch is computed based on the time of this call.
+func (r *eventReporter) startFetchTimer() {
+	r.start = time.Now()
+}
+
+func (r *eventReporter) Done() <-chan struct{} {
+	return r.done
+}
+
+func (r *eventReporter) Event(event common.MapStr) bool {
+	return r.ErrorWith(nil, event)
+}
+
+func (r *eventReporter) Error(err error) bool {
+	return r.ErrorWith(err, nil)
+}
+
+func (r *eventReporter) ErrorWith(err error, meta common.MapStr) bool {
+	timestamp := r.start
+	elapsed := time.Duration(0)
+
+	if !timestamp.IsZero() {
```

Comment:
Minor comment / thought: In `Fetch() []common.MapStr` all events have the exact same time which elapsed. In this new interface the first event will have a lower `rtt` then the last event. Not sure yet if this is a good or bad thing.